extended existing HLT tau-jet correlation module

brallmond opened the pull request:
PR description:

To support a future VBF Single Tau HLT in the scenario that VBF Parking is accepted (see JIRA https://its.cern.ch/jira/browse/CMSLITDPG-1099 and https://its.cern.ch/jira/browse/CMSHLT-2702), a small change to an existing HLT module is necessary. The change extends the module's usage from only handling 2 Jet + 2 Tau events to also handling 2 Jet + 1 Tau events. This change depends on no other PRs and affects code used by one existing HLT path, and would allow the implementation of a future HLT path.

PR validation:

In the course of estimating rate and gain for the proposed VBF Single Tau path, the same metrics were also computed for the VBF DiTau path with the module extension in place and no differences were found in past and current behavior of the path.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34502

- This PR adds an extra 12KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34502/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34502/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34502/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34502/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -68,7 +71,7 @@ void HLTPFDiJetCorrCheckerWithDiTau::produce(edm::StreamID iSId, edm::Event& iEv
   std::set<unsigned int> indices;
 
-  if (pfJets.size() > 1 && taus.size() > 1) {
+  if (pfJets.size() > 1 && taus.size() > 0) {
```
Since the loop is on jet pairs, this `if` statement looks redundant altogether. Doesn't the plugin behave the same without it? If so, please remove it.

brallmond replied:
"As far as I understand, this is because that single use case in the current HLT is preceded by a filter requiring the input taus are at least 2. In general, this update changes the behaviour of the filter. Is that correct?" 
Correct. The behavior of the filter is the same in the case of 2 jets and 2 taus and the proposed change allows the same cross-cleaning/correlation-checking to be performed with 2 jets and 1 tau.

The plugin behaves the same without the if statement, as you pointed out. It has been removed

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -84,17 +87,21 @@ void HLTPFDiJetCorrCheckerWithDiTau::produce(edm::StreamID iSId, edm::Event& iEv
-
-            correctComb = true;
-            break;
+          if (taus.size() == 1) { correctComb = true; }
```
In the case of 1 tau, shouldn't the pT requirement on `iTau1` be considered? (right now it isn't)

brallmond replied:
Yes, thanks for pointing this out. It should be considered and I just added it.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34504

- This PR adds an extra 12KB to repository

cmsbuild commented:
A new Pull Request was created by  @brallmond (Braden Allmond) for master.

It involves the following packages:

- RecoTauTag/HLTProducers (**hlt**)


@cmsbuild, @missirol, @Martin-Grunewald can you please review it and eventually sign? Thanks.
@silviodonato, @missirol, @azotz, @mbluj this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol reviewed (commented):
This update of the filter doesn't fit very well with the name of the plugin ("*WithDiTau"). One option is to add a copy of this plugin with a more suitable name (and we keep the current one in 13_1_X and 13_0_X for backward compatibility). This could be done by just doing
```cpp
DEFINE_FWK_MODULE(HLTPFDiJetCorrCheckerWithDiTau);
using HLTDiPFJetPlusTauFilter = HLTPFDiJetCorrCheckerWithTaus;
DEFINE_FWK_MODULE(HLTDiPFJetPlusTausFilter);
```
Eventually (e.g. 13_2_X), the old plugin would be removed (and replaced at HLT by the new one). Thoughts?

> the same metrics were also computed for the VBF DiTau path with the module extension in place and no differences were found in past and current behavior of the path.

As far as I understand, this is because that single use case in the current HLT is preceded by a filter requiring the input taus are at least 2. In general, this update changes the behaviour of the filter. Is that correct?

brallmond commented:
> This update of the filter doesn't fit very well with the name of the plugin ("*WithDiTau"). One option is to add a copy of this plugin with a more suitable name (and we keep the current one in 13_1_X and 13_0_X for backward compatibility). This could be done by just doing
> 
> ```c++
> DEFINE_FWK_MODULE(HLTPFDiJetCorrCheckerWithDiTau);
> using HLTDiPFJetPlusTauFilter = HLTPFDiJetCorrCheckerWithTaus;
> DEFINE_FWK_MODULE(HLTDiPFJetPlusTausFilter);
> ```
> 
> Eventually (e.g. 13_2_X), the old plugin would be removed (and replaced at HLT by the new one). Thoughts?

Yes, I agree that's a good way to change the filter name in the future and proceed with the change. The code has been added. I think "HLTPFDiJetCorrCheckerWithTaus" above should have been "HLTPFDiJetCorrCheckerWithDiTau" so I changed it in the new commit

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34507

- This PR adds an extra 12KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34507/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34507/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -71,48 +71,50 @@ void HLTPFDiJetCorrCheckerWithDiTau::produce(edm::StreamID iSId, edm::Event& iEv
-          if (reco::deltaR2(taus[iTau1]->p4(), myPFJet1.p4()) < dRmin2_)
-            continue;
-          if (reco::deltaR2(taus[iTau1]->p4(), myPFJet2.p4()) < dRmin2_)
+        if (taus.size() == 1) { 
+          if (taus[iTau1]->pt() < extraTauPtCut_) 
```
Can we move `taus[iTau1]->pt() < extraTauPtCut_` outside this `if`, and remove the 2nd `taus[iTau1]->pt() < extraTauPtCut_` further below in the `else` ?

brallmond replied:
If we move those statements as you suggest then in the case of two taus the cut is applied to both. It is only intended to be applied to the leading tau, hence its inclusion in the ```if``` as well as the ```else``` statement. In the case of two taus, the pt isn't ordered. If they were pt-ordered, we could put ```taus[iTau1]->pt() < extraTauPtCut_``` outside the if statement and remove the check on the 2nd tau entirely.

missirol replied:
Okay, thanks for the clarification.

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -143,3 +145,5 @@ void HLTPFDiJetCorrCheckerWithDiTau::fillDescriptions(edm::ConfigurationDescript
 //
 #include "FWCore/Framework/interface/MakerMacros.h"
 DEFINE_FWK_MODULE(HLTPFDiJetCorrCheckerWithDiTau);
+using HLTDiPFJetPlusTauFilter = HLTPFDiJetCorrCheckerWithDiTau;
+DEFINE_FWK_MODULE(HLTDiPFJetPlusTauFilter);
```
Wouldn't it be clearer as `Taus` (plural) since it can be more than one?

brallmond replied:
Yes, my mistake. I've now added the change.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34508

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #40993 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -68,22 +71,30 @@ void HLTPFDiJetCorrCheckerWithDiTau::produce(edm::StreamID iSId, edm::Event& iEv
+          correctComb = true;
+          break;
+        }
+
+        else {
```
Please remove the empty line between the `if` and the `else` blocks.

brallmond replied:
So, I've never squashed/rebased before in git but I tried the following. 
git rebase -i HEAD~6
and set the first commit (the one at the top) to "pick" and the rest to "squash" in the interactive editor that popped up. 
However, it seems that my local changes were reverted to the first commit (so i should have squashed everything up to the last commit, and then changed that commit message). Somehow I am not able to pull the changes from this PR back (with git pull on the branch that I had been pushing from).

How do you suggest I proceed?

missirol replied:
I would do something like the following (most likely, there are much better ways):
```sh
cmsrel CMSSW_13_1_0_pre1
cd CMSSW_13_1_0_pre1/src
cmsenv
git cms-init
git cms-addpkg RecoTauTag/HLTProducers
git checkout my-cmssw/master-branch-extend-HLTModule-PFDiJetCorrChecker RecoTauTag/HLTProducers/src
git checkout -b master-branch-extend-HLTModule-PFDiJetCorrChecker

## check that the local changes are as intended
## fix manually the empty line as per the last comment

git add -u :/
git commit -m 'extended existing HLT tau-jet correlation module' 
scram b
scram b code-checks
scram b code-format
scram b
git push my-cmssw master-branch-extend-HLTModule-PFDiJetCorrChecker -f
```

brallmond replied:
Thank you so much for the instructions. I think the lines 
```
git add -u :/
git commit -m 'extended existing HLT tau-jet correlation module' 
```
should go after the code-checks/formats, at least that's what I figured out just now since there was a code-formatting error that was fixed locally but not committed before the push. Everything should be resolved now, and also on the backport.

missirol replied:
You are right, my mistake. I edited my previous message for clarity.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34509

- This PR adds an extra 12KB to repository

missirol reviewed (commented):
One more small comment below. Please take this last change as an opportunity to squash the commits into 1, and force-push the changes as 1 commit (to simplify the commit history). That should be all.

cmsbuild commented:
Pull request #40993 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

brallmond committed:
extended existing HLT tau-jet correlation module

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34515

- This PR adds an extra 12KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34515/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34515/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

brallmond committed:
extended existing HLT tau-jet correlation module

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34516

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #40993 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6880f2/31147/summary.html
**COMMIT**: 66f9c28ad44fdfc176100e461d28c5a2cc9e2a3f
**CMSSW**: CMSSW_13_1_X_2023-03-08-1300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40993/31147/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 13 lines to the logs
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3530381
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3530350
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-08-1300+6880f2/55904/triggerResults): no differences found

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -68,22 +71,28 @@ void HLTPFDiJetCorrCheckerWithDiTau::produce(edm::StreamID iSId, edm::Event& iEv
             continue;
 
+          correctComb = true;
+          break;
```
```suggestion
```
The `break` here can be removed, as it's sort-of redundant with the one in L110.

brallmond replied:
Right. I'll also remove the ```break``` on line 106 for the same reason. Both cases are caught by the lines 109 and 110.

missirol replied:
Mh.. the one on L106 is needed to exit the 2nd loop on the tau candidates (which is only in the `else` block and not in the `if` block); I think it makes sense to keep it (i.e. the one on L106).

missirol commented on RecoTauTag/HLTProducers/src/HLTPFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -134,3 +143,5 @@ void HLTPFDiJetCorrCheckerWithDiTau::fillDescriptions(edm::ConfigurationDescript
 //
 #include "FWCore/Framework/interface/MakerMacros.h"
 DEFINE_FWK_MODULE(HLTPFDiJetCorrCheckerWithDiTau);
+using HLTDiPFJetPlusTausFilter = HLTPFDiJetCorrCheckerWithDiTau;
+DEFINE_FWK_MODULE(HLTDiPFJetPlusTausFilter);
```
```suggestion
using HLTDiPFJetPlusTausPFJetProducer = HLTPFDiJetCorrCheckerWithDiTau;
DEFINE_FWK_MODULE(DiPFJetPlusTausPFJetProducer);
```
I have to change my earlier suggestion (sorry..). This is not a good name, as this is an EDProducer (not an EDFilter). Feel free to suggest a better name.

missirol replied:
In general, I would use `*PFJetProducer` (as opposed to `*DiPFJetProducer`), as the output is a list of jets (not a list of jet pairs). Words like "compute", "decorrelate" or "clean" in the plugin name are either too obvious or vague, imho.

`HLTDiPFJetPlusTausPFJetProducer` is shorter (pro). It is more vague (con), but this could also help in the future (pro) if the producer is extended (with extra options, extra cuts, etc). Example: in the future someone might want to use this plugin turning off the Mjj cut (can be done with `mjjMin = -1`), and adding the option to cut on something else, "X(j,j)"; in that scenario, having "Mjj" in the plugin name is not great.

The idea is that this outputs the jets that are part of "good DiJet + Tau(s) candidates". Maybe something like `HLTDiPFJetPlusTausTopologyPFJetProducer` or `HLTDiPFJetPlusTausCandidatePFJetProducer` might be clearer (I tend to favour the latter).

brallmond replied:
I see your point. I agree and think that ```HLTDiPFJetPlusTausCandidatePFJetProducer``` is more future proof. I'll make the change and commit shortly

missirol replied:
Sounds good. You might as well squash again in 1 commit at the next (force) push.

brallmond committed:
extended existing HLT tau-jet correlation module

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40993/34528

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #40993 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
please test

(Didn't squash in 1 commit, but having 3 commits is not a problem, imho.)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6880f2/31161/summary.html
**COMMIT**: 53359747c12b26f5a5ef6ab6b9d16828eacd9ffa
**CMSSW**: CMSSW_13_1_X_2023-03-08-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40993/31161/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 4 lines to the logs
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3530381
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3530358
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-08-2300+6880f2/55916/triggerResults): no differences found

missirol commented:
+hlt

 - extends the logic of an existing HLT filter
 - does not change the results of the current HLT menu
 - adds a copy of the plugin with a clearer name, to be adopted in future HLT menus (to be done in [CMSHLT-2700](https://its.cern.ch/jira/browse/CMSHLT-2700))

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
